README.md
# Ultokengen

Ultokengen is a **Python** program (based on the upLynk API) that when executed, prompts the user for a series of parameters and then compiles a playback URL for DRM-protected content.

Two versions of the program are included that are optimized to run in different environments:

* **native UNIX/Linux (ultokengen2.py)**

* **native Windows (ultokengen2_win.py)**

The programs mints the token based on a number of parameters and appends them to the end of the playback URL.

The expiration time of the token hard-coded at current UTC time + 100 year (3,136,320,000 seconds).

## Installation

Browse to your desired install sub-directory and run the following code:

* For **UNIX/Linux**

```PYTHON
curl "https://raw.githubusercontent.com/jj358mhz/Ultokengen/master/ultokengen2.py" -o ultokengen2.py
```

* For **Windows**

```PYTHON
curl "https://raw.githubusercontent.com/jj358mhz/Ultokengen/master/ultokengen2_win.py" -o ultokengen2_win.py
```

## What you will be prompted for is:
* The API key (found in upLynk CMS)
* The protected playback URL
* Asset type


CHANGELOG.md
# CHANGELOG


## Version 1.1.4 - (13-Sept-2017)
* UPDATED: Live event prompt for content type

## Version 1.1.3 - (22-Aug-2016)
* ADDED: Rays and ad parameter options
* UPDATED: Minor code text changes

## Version 1.1.2 - (21-May-2015)
* UPDATED: Minor code text changes

## Version 1.1.1 - (20-May-2015)
* FIXED: GUID extraction code

## Version 1.1 - (18-May-2015)
* ADDED: Program will generate an XML-compliant URL
* ADDED: Program will auto-extract the asset GUID from the URL
* CHANGED: Extended the token life 100 years from the time the token is minted

## Version 1.02 - (28-Apr-2014)
* ADDED: New file (ultokengen_win.py). Program customized to run on Windows platforms
* FIXED: Eliminated redundant lines of code

## Version 1.01 - (23-Apr-2014)
* CHANGED: Added an expiry of 1 year (31,557,600 seconds) from the time the token is minted

## Version 1.00 - (22-Apr-2014)
* NEW: ultokengen.py


token.py
import hashlib, time, hmac, urllib

# inputs
ip = '72.27.179.200' # the user's IP address; normally obtained from the server itself
apiKey = 'WxQpQhHFmE4hTWA4TGLu6rYeNuKgYrWwlCLmSKRb' # from the CMS UI

# combine all of the parameters except the signature
queryStr = urllib.urlencode(dict(
    exp = int(time.time()) + 120, # expire 120 seconds from now
    ct = 'a', # an asset
    cid = 'ea10fa402fec4bbe996019a0827e6c38', # the asset's ID
    iph = hashlib.sha256(ip).hexdigest(), # the user's IP address hash
    rays = 'abc', # customization parameter
))

# compute the signature and add it to the *end*
sig = hmac.new(apiKey, queryStr, hashlib.sha256).hexdigest()
queryStr = queryStr + '&sig=' + sig

# The token would then be added to a playback URL, e.g.
url = 'http://content.uplynk.com/ea10fa402fec4bbe996019a0827e6c38.m3u8'
url = url + '?' + queryStr


ultokengen.py
#!/usr/bin/env python

# Version 1.02 (28-Apr-2014)

import hashlib
import time
import hmac
import urllib

# inputs
ip = raw_input('Enter your IP address : ')  # the user's IP address
apiKey = raw_input('Enter your API key : ')  # from the CMS UI
pbURL = raw_input('Enter the Playback URL : ')  # the playback URL from the upLynk CMS
ct = raw_input('Enter the content type, a for asset or c for live channel : ')  # the content type
cid = raw_input('Enter the GUID : ')  # the content ID from the upLynk CMS
rays = raw_input('Enter the customization parameters, abc : ')  # customization parameter

# combine all of the parameters except the signature
queryStr = urllib.urlencode(dict(
    exp=int(time.time()) + 31557600,  # expire 1 year from now
    ct=ct,  # an asset
    cid=cid,  # the asset's ID
    iph=hashlib.sha256(ip).hexdigest(),  # the user's IP address hash
    rays=rays,  # customization parameter
))

# compute the signature and add it to the *end*
sig = hmac.new(apiKey, queryStr, hashlib.sha256).hexdigest()
queryStr = queryStr + '&sig=' + sig

# The token would then be added to a playback URL, e.g.
url = pbURL + '?' + queryStr
print (url)


ultokengen2.py
# ultokengen2
# Version 1.1.4

'''
This program will tokenize a DRM-protected URL
provided a valid API key is entered when prompted.
It will present both a standard tokenized playback URL and
one that is XML-compliant for XML feed readers.
'''

import hashlib
import time
import hmac
import urllib
import random

# inputs
apiKey = raw_input('Enter your API key : ')  # from the CMS UI
url = raw_input('Enter the Playback URL : ')  # the playback URL from the upLynk CMS
ct = raw_input('Enter the content type, "a" for asset or "c" for live channel or "e" for live event : ')  # the content type
cid = url.split('/')[-1][:-5]  # extracts the GUID
print ("The extracted GUID is: "+cid)
# rays = raw_input('Enter the customization parameters, abc : ')  # customization parameter, uncomment if needed
# ad = raw_input('Enter the ad server name : ')  # customization parameter, uncomment if needed

# combine all of the parameters except the signature
queryStr = urllib.urlencode(dict(
    tc='0',  # token check algorithm version; 0 = no check, 1 = check
    exp=int(time.time()) + 3136320000,  # expire 100 years from now
    rn=str(random.randint(0, 2**32)),  # random number
    ct=ct,  # an asset
    cid=cid,  # the asset's ID
    # rays=rays,  # customization parameter, uncomment if needed
    # ad=ad,  # customization parameter, uncomment if needed
))

# compute the signature and add it to the *end*
sig = hmac.new(apiKey, queryStr, hashlib.sha256).hexdigest()
queryStr = queryStr + '&sig=' + sig

# the token would then be added to a playback URL, e.g.
turl = url + '?' + queryStr
print ('Standard URL: '+turl)

# the following code provides a XML-compliant URL
new_string = turl.replace('&', '&#38;')
final_url = new_string.replace('?', '&#63;')
print ('XML-compliant URL: '+final_url)


ultokengen2_win.py
#!/usr/bin/env python
# Version 1.1.4
# program customized to run on Windows platforms

'''
This program will tokenize a DRM-protected URL
provided a valid API key is entered when prompted.
It will present both a standard tokenized playback URL and
one that is XML-compliant for XML feed readers.
'''

import hashlib, time, hmac, urllib.parse, random

# inputs
apiKey = raw_input('Enter your API key : ')  # from the CMS UI
url = raw_input('Enter the Playback URL : ')  # the playback URL from the upLynk CMS
ct = raw_input('Enter the content type, "a" for asset or "c" for live channel or "e" for live event : ')  # the content type
cid = url.split('/')[-1][:-5]  # extracts the GUID
print ("The extracted GUID is: "+cid)
# rays = raw_input('Enter the customization parameters, abc : ')  # customization parameter, uncomment if needed
# ad = raw_input('Enter the ad server name : ')  # customization parameter, uncomment if needed

# combine all of the parameters except the signature
queryStr = urllib.urlencode(dict(
    tc='0',  # token check algorithm version
    exp=int(time.time()) + 3136320000,  # expire 100 years from now
    rn=str(random.randint(0, 2**32)),  # random number
    ct=ct,  # an asset
    cid=cid,  # the asset's ID
    # rays=rays,  # customization parameter, uncomment if needed
    # ad=ad,  # customization parameter, uncomment if needed
))

# compute the signature and add it to the *end*
sig = hmac.new(apiKey, queryStr, hashlib.sha256).hexdigest()
queryStr = queryStr + '&sig=' + sig

# the token would then be added to a playback URL, e.g.
turl = url + '?' + queryStr
print ('Standard URL: '+turl)

# the following code provides a XML-compliant URL
new_string = turl.replace('&', '&#38;')
final_url = new_string.replace('?', '&#63;')
print ('XML-compliant URL: '+final_url)


ultokengen_win.py
#!/usr/bin/env python

# program customized to run on Windows platforms
# Version 1.02 (28-Apr-2014)

import hashlib, time, hmac, urllib.parse

# inputs
ip = input('Enter your IP address : ') # the user's IP address; normally obtained from the server itself
print ("Hi %s, I know where you live!" % ip);
apiKey = input('Enter your API key : ') # from the CMS UI
print ("Shhhhh, be sure to keep %s secret!" % apiKey);
pbURL = input('Enter the Playback URL : ') # the playback URL from the upLynk CMS
print ("%s, interesting URL" % pbURL);
ct = input('Enter the content type, a for asset or c for live channel : ') # the content type
print ("%s. Cool" % ct);
cid = input('Enter the GUID : ') # the content ID from the upLynk CMS
print ("Wow, %s is a big number!" % cid);
rays = input('Enter the customization parameters, abc : ') # customization parameter
print ("With this %s, I can begin to compute" % rays);

# combine all of the parameters except the signature
queryStr = urllib.parse.urlencode(dict(
    exp = int(time.time()) + 31557600, # expire 1 year from now
    ct = ct, # an asset
    cid = cid, # the asset's ID
    iph = hashlib.sha256(ip.encode()).hexdigest(), # the user's IP address hash
    rays = rays, # customization parameter
))

# compute the signature and add it to the *end*
sig = hmac.new(apiKey.encode(),queryStr.encode(),hashlib.sha256).hexdigest()
queryStr = queryStr + '&sig=' + sig

# The token would then be added to a playback URL, e.g.
url = pbURL + '?' + queryStr
print (url)
input('Press ENTER to exit')
